### ISSUE TITLE
Feature Request: General Purpose Metrics Callback

### ISSUE THREAD

**Issue by PhilipMay** [tensorflow/tensorflow#36799]
# Background
I am coming from tensorflow addons and have been send here by @pavithrasv and @seanpmorgan to start this discussion. The root of this is that we have great problems to implement a clean and bugfree F1 metric for keras. See here: https://github.com/tensorflow/addons/issues/825

# Feature Request Description
For keras / tensorflow it is possible to implement custom metrics. These metrics are basicaly keras layers and have to be implemented with tensorflow itself. For many reasons it is not easy to implement certain metrics. This is the reason why some metrics like F1 are still not available in TF and have a history of ugly bugs. This is why I think that it would be good to implement a callback to plug in any metric without having to use TF. So you could plug in all sklearn metrics you want to use.
This is the case with other tools like LightGBM and XGBoost. See `feval` parameter of `lightgbm.train`: https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html#lightgbm.train

# First Idea of Implementation
A first implementation version could look like this (but has some limitations - see below):

``` python
class MetricsCallback(keras.callbacks.Callback):
    def __init__(self, val_data, val_labels, fmetrics):
        super().__init__()
        self.val_data = val_data
        self.val_labels = val_labels
        self.fmetrics = fmetrics
        # TODO: more checks

    def on_epoch_end(self, epoch, logs=None):
        logs = logs or {}
        predict_results = self.model.predict(self.val_data)

        for fm in self.fmetrics:
            metric_name, value = fm(self.val_labels, predict_results)
            logs[metric_name] = value
```

## Limitations
You have to give validation data to this callback and not to the fit function of keras. Otherwise the validation data will be predicted twice - once from keras fit and once from this callback. 

And if you want to use the EarlyStopping callback you have to provide this callback before the EarlyStopping callback.

# Next Steps
What are the next steps? I would like to start a discussion on how to implement the first version in a way that integrates with keras and does not feel like a hack.

**Comment by tillmo** [comment]
As I wrote in https://github.com/tensorflow/addons/issues/825, it seems that `keras.metrics.Precision(name='precision')` and `keras.metrics.Recall(name='recall')` already solve the batch problem. So keras would only need to add the obvious F1 computation from these values. Then there is no need to add all this to tensorflow.

**Comment by gabrieldemarmiesse** [comment]
If one wants to be able to use numpy in a `Metric`, I believe that's possible. A metric doesn't require any gradient and you just have to execute `result` and `update_states` in eager mode. I mean, in eager mode you can do pretty much anything you want anyway, so I fail to see how using numpy would be a problem.

**Comment by naji-s** [comment]
@gabrieldemarmiesse I don't think the F1 implemented in Addons is bug-free. It keeps complaining about the ranks because it forces the output (even in binary case) to be one-hot vector as pointed out [here](https://github.com/tensorflow/addons/issues/746#issuecomment-562963134). I do think this needs to change.

**Comment by John-8704** [comment]
> If one wants to be able to use numpy in a `Metric`, I believe that's possible. A metric doesn't require any gradient and you just have to execute `result` and `update_states` in eager mode. I mean, in eager mode you can do pretty much anything you want anyway, so I fail to see how using numpy would be a problem.
> 
> > The root of this is that we have great problems to implement a clean and bugfree F1 metric for keras.
> 
> TensorFlow addons have (and had since quite some time) a clean and bugfree F1 metric implementation for tf.keras. Not for multi-backend keras though, but this is fine since multi-backend keras is deprecated. The issue mentioned here: [tensorflow/addons#825](https://github.com/tensorflow/addons/issues/825) was just a misunderstanding concerning the compatibility between multi-backend keras and TF Addons. As such, it's not a valid bug report.

Does the f1-score metric implementation in tf-addons calculate score per batch wise (or) on the full validation set? It wasn't mentioned clearly.

**Comment by gabrieldemarmiesse** [comment]
It's computing the score on the full validation set to get more meaningful results :)

**Comment by jvishnuvardhan** [comment]
@PhilipMay Is there any actionable item like raising PR? 

Please note that Keras development moved to another repository to focus entirely on only keras. Could you please repost this issue on [keras-team/keras repo](https://github.com/keras-team/keras/issues). Thanks!
